internal/worker/sshserver/server.go

```diff
@@ -137,11 +132,10 @@ func NewServerWorker(config ServerWorkerConfig) (worker.Worker, error) {
 		// Keep the listener and the server alive until the tomb is killed.
 		<-s.tomb.Dying()
 
-		// Close the listener, this prevents a race in the test.
-		if err := listener.Close(); err != nil {
-			s.config.Logger.Errorf("failed to close listener: %v", err)
-		}
-
+		// Wait on the closeAllowed to indicate when we can stop the SSH server.
+		// This ensures we don't face a race condition, closing the server
+		// too quickly after it was started, see listener.go for more details.
+		<-closeAllowed
```

Comment:
This looks quite scary, but it is ok. The server cannot exit without at least calling the defered close on the listener which will close this channel and allow this to proceed.